packages/venia-ui/lib/RootComponents/Product/product.js

```diff
@@ -14,49 +15,35 @@ import getUrlKey from '../../util/getUrlKey';
  * TODO: Replace with a single product query when possible.
  */
 import GET_PRODUCT_DETAIL from '../../queries/getProductDetail.graphql';
+import PRODUCT_DETAILS_FRAGMENT from '../../fragments/productDetails.graphql';
 
 const Product = () => {
     useEffect(() => {
         window.scrollTo(0, 0);
     }, []);
 
-    const { loading, error, data } = useQuery(GET_PRODUCT_DETAIL, {
-        variables: {
-            onServer: false,
-            urlKey: getUrlKey()
-        }
+    const talonProps = useProduct({
+        fragment: PRODUCT_DETAILS_FRAGMENT,
+        query: GET_PRODUCT_DETAIL,
+        urlKey: getUrlKey()
     });
 
-    // Memoize the result from the query to avoid unnecessary rerenders.
-    const product = useMemo(() => {
-        if (!data) {
-            return;
-        }
-        const product = data.productDetail.items[0];
-        // map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to
-        // maintain backwards compatibility
-        const { description } = product;
-        return {
-            ...product,
-            description:
-                typeof description === 'object' ? description.html : description
-        };
-    }, [data]);
+    const { error, loading, product } = talonProps;
+
+    if (product) {
+        // Note: STORE_NAME is injected by Webpack at build time.
+        return (
+            <Fragment>
+                <Title>{`${product.name} - ${STORE_NAME}`}</Title>
+                <ProductFullDetail product={product} />
+            </Fragment>
+        );
+    }
 
     if (loading) return fullPageLoadingIndicator;
     if (error) return <div>Data Fetch Error</div>;
 
-    if (!product) {
-        return <ErrorView outOfStock={true} />;
-    }
-
-    // Note: STORE_NAME is injected by Webpack at build time.
-    return (
-        <Fragment>
-            <Title>{`${product.name} - ${STORE_NAME}`}</Title>
-            <ProductFullDetail product={product} />
-        </Fragment>
-    );
+    return <ErrorView outOfStock={true} />;
```

Comment:
I think the last `return` should be the "happy path" render. In this case we could leave it as it was right?

Reply:
Oh yeah, thanks for pointing this out too. This is a definite departure from how we usually render, but `useLazyQuery` sets `loading` to `true` while the "background refresh" request is in flight.

So even though it had a cached product it could use, it'd show the full page loading indicator until the background refresh was complete, which defeated the entire purpose 😜

Reply:
Could we just change that line to `if (loading && !data) return fullPageLoadingIndicator;`?

Reply:
I demand that you not use ErrorView until it throws errors in developer mode.

Reply:
Yes, this is also a good time to remind everyone that `ErrorView`

- was designed to handle error types coming from `MagentoRouteHandler` only
- should not have had an `outOfStock` prop added to it
- should not be rendered here

If the custom error view for this component happens to have the same presentation as the one we use for routing failure, its CSS can use `composes` to share styles. 👍

Reply:
Moved happy path to last return for consistency in https://github.com/magento/pwa-studio/pull/1935/commits/55253609a051a9e2e393ff5a94ca3fc489250a4b.

Reply:
I did not add `ErrorView` here, do y'all want me to refactor it out as part of this PR?

Reply:
Nevermind, it was easy: https://github.com/magento/pwa-studio/pull/1935/commits/2413375d7c51ac7004a22f699a6c4cfa1a8987e4.